core/src/main/scala/org/apache/spark/Partitioner.scala

```diff
@@ -57,7 +60,8 @@ object Partitioner {
   def defaultPartitioner(rdd: RDD[_], others: RDD[_]*): Partitioner = {
     val rdds = (Seq(rdd) ++ others)
     val hasPartitioner = rdds.filter(_.partitioner.exists(_.numPartitions > 0))
-    if (hasPartitioner.nonEmpty) {
+    if (hasPartitioner.nonEmpty
+      && isEligiblePartitioner(hasPartitioner.maxBy(_.partitions.length), rdds)) {
```

Comment:
`hasPartitioner.maxBy(_.partitions.length)` is used repeatedly, pull that into a variable ?